pkg/codec/ffmpeg/ffmpeg.go

```diff
@@ -0,0 +1,452 @@
+// Package ffmpeg brings libavcodec's encoding capabilities to mediadevices.
+// This package requires ffmpeg headers and libraries to be built.
+// For more information, see https://github.com/asticode/go-astiav?tab=readme-ov-file#install-ffmpeg-from-source.
+//
+// Currently, only nvenc, x264, vaapi are implemented, but extending this to other ffmpeg supported codecs should
+// be simple.
+package ffmpeg
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/asticode/go-astiav"
+	"github.com/pion/mediadevices/pkg/codec"
+	"github.com/pion/mediadevices/pkg/io/video"
+	"github.com/pion/mediadevices/pkg/prop"
+)
+
+type hardwareEncoder struct {
+	codecCtx       *astiav.CodecContext
+	hwFramesCtx    *astiav.HardwareFramesContext
+	frame          *astiav.Frame
+	hwFrame        *astiav.Frame
+	packet         *astiav.Packet
+	width          int
+	height         int
+	r              video.Reader
+	nextIsKeyFrame bool
+
+	mu     sync.Mutex
+	closed bool
+}
+
+type softwareEncoder struct {
+	codec          *astiav.Codec
+	codecCtx       *astiav.CodecContext
+	frame          *astiav.Frame
+	packet         *astiav.Packet
+	width          int
+	height         int
+	r              video.Reader
+	nextIsKeyFrame bool
+
+	mu     sync.Mutex
+	closed bool
+}
```

Comment:
Maybe we can 
```suggestion
type baseEncoder struct {
	codecCtx       *astiav.CodecContext
	frame          *astiav.Frame
	packet         *astiav.Packet
	width          int
	height         int
	r              video.Reader
	nextIsKeyFrame bool

	mu     sync.Mutex
	closed bool
}

type hardwareEncoder struct {
	baseEncoder
  
	hwFramesCtx    *astiav.HardwareFramesContext
	hwFrame        *astiav.Frame
}

type softwareEncoder struct {
	baseEncoder

	codec          *astiav.Codec
}
```
to reuse the common members and functions between `hardwareEncoder` and `softwareEncoder`